docs/389ds/design/worker-threads.md

```diff
@@ -0,0 +1,345 @@
+      - Fill the pblock with conn, op
+      - return CONN_FOUND_WORK_TO_DO
+
+### Dynamic change of the number of threads
+
+The number of threads may be changed dynamically by changing the
+ nsslapd-threadnumber attribute in cn=config (when calling config_set_threadnumber). 
+Restart is no more needed.
+
+When changing the number threads:
+
+- if adding threads:
+  - Get global mutex
+  - Compute first new thread index:
+      (size of working_threads list + size of busy_threads list) +1
+  - For each missing threads:
+    - Alloc the thread context struct
+    - Init its mutex and condition variable
+    - Store the thread index in context then increment it
+    - Create the thread (as joinable)
+    - Store the thread id in the context
+  - Release the global mutex
+- else if removing threads:
+  - create a local empty "closing" list
+  - Walk waiting_threads list:
+    - if thread index is higher than wanted number of thread
+      - Move thread from waiting_threads list to closing list
+  - Walk busy_threads list:
+    - if thread index is higher than wanted number of thread
+      - Move thread from busy_threads list to closing list
+  - Release the global mutex
+  - Walk closing list:
+    - Get thread mutex
+    - Set thread closing flags
+    - Wake up the thread
+    - Release thread mutex
+  - Walk closing list:
+    - Join the thread
+    - Unlink thread from closing list
+    - Free thread context
+
+### Monitoring
+
+The following data should be displayed:
+
+| Data                        | Attribute name | Description                                                   |
+| --------------------------- | -------------- | ------------------------------------------------------------- |
+| waiting_thread size         | waitingthreads | Number of worker threads that are waiting for jobs            |
+| busy_thread size            | busythreads    | Number of worker threads that are processing on  jobs         |
+| busy_thread high water mark | maxbusythreads | Highest number of worker threads that are processing on  jobs |
+| waiting_job size            | waitingjobs    | Size of the waiting job queue                                 |
+| waiting_job high water mark | maxwaitingjobs | Highest size of the waiting job queue                         |
+
+### Test
+
+An automatic test can be done with the following steps:
+
+- create an instance
+- get the monitoring results
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one 
+- increase the number of threads
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one
+- decrease the number of threads
+- check that waitingthreads+busythreads == configured number of threads
+- run pstack of server and check that number of connection_threadmain is the expected one
+
```

Comment:
For support aspect, you may test setting 1000 workers and check how long it takes to get a pstack.
Then doing the same job while running a searchrate/ldclt load

Reply:
ok but that is is outside the automatic test scope ...
Will add a new section for other manual tests

Reply:
Done (see just above this comment)